### PR TITLE
Add core Biscuit tools

### PR DESCRIPTION
**Pull request by njspix**
Add core Biscuit [huishenlab.github.io](url) modules (`index`, `align`, and `pileup`). Planning to add more tools (`bsconv`, `epiread`) later, as well as subworkflows.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

**Comment by njspix** [comment]
I've added almost all the biscuit subcommands, as well as a 'blaster' subcommand that incorporates duplicate marking, etc. in one step.

**Comment by jfy133** [comment]
@njspix I think you're going to scare off a lot of people from reviewing this ;)

image

please try and revert the file permissions changes

**Comment by njspix** [comment]
Ahh, so sorry about that! Thanks for saying something. I'll try to get that cleaned up...

**Comment by njspix** [comment]
@jfy133 Thanks so much for working on this! I added the `when` blocks to the appropriate files. 

Sorry about the large PR - I'll keep that in mind in the future.

File: modules/biscuit/align/main.nf
Comment (sateeshperi) [inline modules/biscuit/align/main.nf:26]: wow! is this cpu setting specific for biscuit?

Comment (njspix) [reply modules/biscuit/align/main.nf:26]: Yes... The goal is to assign the bulk of the available CPUs to the aligner. In cases where there's a lot of CPUs available though it's been my experience that it helps to pass samtools a few extra cores as well. This hopefully allows for that while avoiding specifying more cores than are actually available...

Looking at it again dividing by 1.11 seems a bit more obtuse than the equivalent 'task.cpus*0.9', so maybe I should change that to make it clearer.

Comment (njspix) [reply modules/biscuit/align/main.nf:26]: The alternatives I've seen are 

1. Pass `task.cpus` to both aligner and samtools (seems reckless when e.g. 20 cores are passed but then upto 40 are 'expected' by the tools, even if there's no way samtools would use 40 cores)
2. Pass `task.cpus` to just the aligner. This doesn't seem quite optimal as with a lot of cores, the aligner could 'pile up' reads in the pipe (producing alignments faster than samtools can consume them), degrading overall performance and possibly memory consumption.

Comment (njspix) [reply modules/biscuit/align/main.nf:26]: Happy to change if you think am alternative approach would be better

Comment (jfy133) [reply modules/biscuit/align/main.nf:26]: I think `task.cpus * 0.9` would be fine here :+1:

File: modules/biscuit/align/main.nf
Comment (jfy133) [inline modules/biscuit/align/main.nf:25]:
```suggestion
```

If this is required, please include it in your input tuple. Otherwise this should be removed as this can be specified by a pipeline developer via `ext.args` in `modules.conf`

File: modules/biscuit/align/main.nf
Comment (jfy133) [inline modules/biscuit/align/main.nf:34]:
```suggestion
```

File: modules/biscuit/biscuitblaster/main.nf
Comment (jfy133) [inline modules/biscuit/biscuitblaster/main.nf:26]: I would suggest just calling these as `args` `args1` etc. as ths is more consistent with other modules (I guess as a biscuit-module user this would be more readable, for people new to it or other non-biscuit experts looking at the module later it would be better to follow the general trend)

File: modules/biscuit/biscuitblaster/main.nf
Comment (jfy133) [inline modules/biscuit/biscuitblaster/main.nf:27]: Should be removed ormoved to `input tuple`, see comment in align

File: modules/biscuit/biscuitblaster/main.nf
Comment (jfy133) [inline modules/biscuit/biscuitblaster/main.nf:28]:
```suggestion
    def biscuit_cpus = (int) Math.max(Math.floor(task.cpus/1.05),1)
```
see above in `align`

File: modules/biscuit/biscuitblaster/meta.yml
Comment (jfy133) [inline modules/biscuit/biscuitblaster/meta.yml:55]:
```suggestion
      description: Biscuit genome index directory
```

output from ...? I say this as you specify a very highly specific pattern for the input here

File: modules/biscuit/bsconv/main.nf
Comment (jfy133) [inline modules/biscuit/bsconv/main.nf:15]: Should just be `*.bam` unless the tool itself writes it like so

see last bullet of this: https://nf-co.re/developers/adding_modules#naming-conventions

If you're worried about filename conflicts

File: modules/biscuit/epiread/main.nf
Comment (jfy133) [inline modules/biscuit/epiread/main.nf:33]:
```suggestion
```

File: modules/biscuit/epiread/main.nf
Comment (jfy133) [inline modules/biscuit/epiread/main.nf:49]:
```suggestion
        -c > ${prefix}.epiread.bed.gz
```

see bsconv

File: modules/biscuit/index/main.nf
Comment (jfy133) [inline modules/biscuit/index/main.nf:22]:
```suggestion
```

File: modules/biscuit/pileup/main.nf
Comment (jfy133) [inline modules/biscuit/pileup/main.nf:17]:
```suggestion
    tuple val(meta), path("*.vcf.gz"), emit: vcf
    path "versions.yml"              , emit: versions

```

File: modules/biscuit/pileupsomatic/main.nf
Comment (jfy133) [inline modules/biscuit/pileupsomatic/main.nf:27]:
```suggestion
```

File: modules/biscuit/pileupsomatic/main.nf
Comment (jfy133) [inline modules/biscuit/pileupsomatic/main.nf:35]: Please confirm this is required for this module, and cannot be supplied by a user/pipeline dev via `modules.conf`

File: modules/biscuit/pileupsomatic/main.nf
Comment (jfy133) [inline modules/biscuit/pileupsomatic/main.nf:11]: I see that the command is just `biscuit pileup`, can you make it more generic by just passing optional input for `tumor_ba*` etc?

so insert the `tumor_*` files only if they are _not_ an empty list `[]` via ` def` in the script block?

File: modules/biscuit/qc/main.nf
Comment (jfy133) [inline modules/biscuit/qc/main.nf:25]:
```suggestion
```

File: modules/biscuit/vcf2bed/main.nf
Comment (jfy133) [inline modules/biscuit/vcf2bed/main.nf:24]:
```suggestion
```

File: modules/biscuit/vcf2bed/main.nf
Comment (jfy133) [inline modules/biscuit/vcf2bed/main.nf:36]:
```suggestion
        biscuit: \$(echo \$(biscuit version 2>&1) | sed 's/^.*BISCUIT Version: //; s/Using.*\$//')
```

File: modules/biscuit/vcf2bed/meta.yml
Comment (jfy133) [inline modules/biscuit/vcf2bed/meta.yml:41]:
```suggestion
      pattern: "*.{bed.gz}"
```

File: tests/modules/biscuit/epiread/main.nf
Comment (jfy133) [inline tests/modules/biscuit/epiread/main.nf:42]: Missing URL suffix?

File: tests/modules/biscuit/index/test.yml
Comment (jfy133) [inline tests/modules/biscuit/index/test.yml:23]:
```suggestion
```

File: tests/modules/biscuit/mergecg/main.nf
Comment (jfy133) [inline tests/modules/biscuit/mergecg/main.nf:12]: Missing URL suffix?

File: tests/config/nextflow.config
Comment (jfy133) [inline tests/config/nextflow.config:31]: Reason for this?

**Review by jfy133 — CHANGES_REQUESTED** [review]
General comments across multiple modules (see specific comments for more info)
- keep consistent `biscuit_args` vs `args` `args2` etc (the latter is preferred)
- go with the simpler maths for the CPUs
- Remove `$read_groups` being taken from meta (pipeline devs can do that with `args`), unless it's mandatory in which case add it to the input tuple 
- output filenames should always just be `${prefix}.<suffix>`, see comment on bsconv for workaround if you fear filename clashes

**Comment by sateeshperi** [comment]
Also consider updating the `documentation` links in each `meta.yml` to tools specific documentation page

File: modules/biscuit/biscuitblaster/meta.yml
Comment (njspix) [reply modules/biscuit/biscuitblaster/meta.yml:55]: Updated...

File: modules/biscuit/pileupsomatic/main.nf
Comment (njspix) [reply modules/biscuit/pileupsomatic/main.nf:11]: I think this could be made to work but it might get pretty ugly pretty fast. If the `-S` flag is set, we need exactly one `tumor_bam` (following the `-T` flag) and exactly one `normal_bam` (following the `-I` flag). These use cases and input requirements seemed different enough to me to split into separate tools. Strelka does similarly.

File: tests/config/nextflow.config
Comment (njspix) [reply tests/config/nextflow.config:31]: Can't recall. Reverted.

**Comment by njspix** [comment]
@jfy133 @sateeshperi thanks so much for all your input and for taking the time to go through this. I think I've addressed everything, would welcome further discussion on what you think the best way forward with `pileup` is.

**Comment by sateeshperi** [comment]
you want to commit the changes for capturing software versions? I don't fully know why its better the way i proposed but, i seen in other PRs that it is

**Comment by njspix** [comment]
@sateeshperi I'm sorry, I'm not seeing the changes you mentioned. I'm still learning Git and GitHub so I'm probably just not looking in the right spot?

**Comment by sateeshperi** [comment]
my bad! i didn't submit my review....see changes requested now

File: modules/biscuit/bsconv/main.nf
Comment (sateeshperi) [inline modules/biscuit/bsconv/main.nf:15]:
```suggestion
    tuple val(meta), path("*.bam")       , emit: bsconv_bam
```

**Review by sateeshperi — CHANGES_REQUESTED** [review]

**Comment by njspix** [comment]
Should be all committed now - let me know if you see anything I missed! Thanks again.

**Review by sateeshperi — CHANGES_REQUESTED** [review]
getting there

File: modules/biscuit/epiread/main.nf
Comment (njspix) [reply modules/biscuit/epiread/main.nf:45]: It can be, but output file would not be properly formatted if these were specified incorrectly. I was viewing this kind of like a gzip command...

**Comment by njspix** [comment]
Ok, I think it's all updated now... for some reason I thought I committed the `task.cpus` changes but they didn't show up...

File: tests/modules/biscuit/vcf2bed/main.nf
Comment (sateeshperi) [inline tests/modules/biscuit/vcf2bed/main.nf:11]: could you use vcf files from `nf-core/test-datasets` ? @njspix

Comment (njspix) [reply tests/modules/biscuit/vcf2bed/main.nf]: No, these .vcf files have methylation information in them specific to this tool. I just updated this with a GitHub path so tests should be able to run now...

**Review by sateeshperi — CHANGES_REQUESTED** [review]
few more

**Comment by njspix** [comment]
I updated the test.ymls and confirmed that all tests are passing locally.

File: modules/biscuit/bsconv/main.nf
Comment (sateeshperi) [inline modules/biscuit/bsconv/main.nf:16]:
```suggestion
    path "versions.yml"           , emit: versions
```

File: modules/biscuit/epiread/main.nf
Comment (sateeshperi) [inline modules/biscuit/epiread/main.nf:16]:
```suggestion
    path "versions.yml"              , emit: versions
```

File: modules/biscuit/index/main.nf
Comment (sateeshperi) [inline modules/biscuit/index/main.nf:15]:
```suggestion
    path "versions.yml"      , emit: versions
```

File: modules/biscuit/pileupsomatic/main.nf
Comment (sateeshperi) [inline modules/biscuit/pileupsomatic/main.nf:16]:
```suggestion
    path "versions.yml"                      , emit: versions
```

**Review by sateeshperi — CHANGES_REQUESTED** [review]
format fix

**Comment by njspix** [comment]
Updated tests for biscuit/align (sorry about the version reporting code, I really don't understand why you keep having to commit that!)

**Comment by sateeshperi** [comment]
those were fixed....these were small formatting fixes...trivial

**Comment by sateeshperi** [comment]
nice! lets wait for the pileup final review. that was a handful! I need a biscuit now :smile:

**Comment by njspix** [comment]
Thanks so much for all your help! Help yourself... https://huishenlab.github.io/biscuit/biscuitblaster/#biscuitblaster-representations

**Review by sateeshperi — APPROVED** [review]

**Review by jfy133 — CHANGES_REQUESTED** [review]
A few more left over things, and a possible sugggestion for the somatic mode.

If there already modules with this `somatic` mode, then I guess you can do it.

I generally don't like this though as I've had fights with @nf-core/sarek about this before e.g. with freebayes (some modules being made too human/pipeline specific ;)).

File: tests/modules/biscuit/epiread/main.nf
Comment (njspix) [reply tests/modules/biscuit/epiread/main.nf:42]: Yes, [here](https://github.com/nf-core/test-datasets/pull/452)

File: modules/biscuit/qc/main.nf
Comment (sateeshperi) [reply modules/biscuit/qc/main.nf:30]: $args on L29

Comment (njspix) [reply modules/biscuit/qc/main.nf:30]: should be fixed now, with the newline slashes (🤦 )

**Comment by njspix** [comment]
Ok, I think everything should be working again. The approach @jfy133 suggested for pileup seems to work nicely

**Comment by sateeshperi** [comment]
almost there

File: tests/modules/biscuit/epiread/main.nf
Comment (jfy133) [reply tests/modules/biscuit/epiread/main.nf:42]: Merged!

File: modules/biscuit/pileup/main.nf
Comment (jfy133) [inline modules/biscuit/pileup/main.nf:29]:
```suggestion
    input = ( tumor_bam == [] ) ? "${normal_bams}" : "-S -T ${tumor_bam} -I ${normal_bams}"
```

**Review by jfy133 — APPROVED** [review]
LGTM now, remaining things:

1. I merged the test-datasets PR so please update the URLs in your tests,
2. check the CI failures then we are GTG

**Comment by njspix** [comment]
Ok, changed the paths (thanks for the merge!) and fixed the offending yml. Please let me know if anything else needs changing!

File: modules/biscuit/pileup/meta.yml
Comment (jfy133) [inline modules/biscuit/pileup/meta.yml:40]:
```suggestion
      description: |
```

**Comment by jfy133** [comment]
@njspix I fixed the last failing CI test, feel free to `squash and merge`!

File: modules/biscuit/pileup/main.nf
Comment (sateeshperi) [reply modules/biscuit/pileup/main.nf:29]: last one here @njspix

**Comment by njspix** [comment]
Hey @jfy133 and @sateeshperi thanks for all your work on this! I got a few emails over the weekend but now looking at GitHub I'm not sure I see any more changes that need to be made - is there anything else I can do?

**Comment by jfy133** [comment]
Nope - wait for tests to complete and you can merge :)

**Comment by njspix** [comment]
Thanks so much! I don't think I can -
Screen Shot 2022-03-14 at 9 33 13 AM

**Comment by njspix** [comment]
🥳 🎉 Thanks so much!!

**Pull request merged**
